### ISSUE TITLE
Clarify what should happen when a resource is unable to run a test

### ISSUE THREAD

**Issue by clintoncwolfe** [inspec/inspec#3174]
### 🎛 Description

🙋 feature request

When a resource is unable to perform a test, what should the outcome of the test be?

Examples of being unable to perform a test:

* `http` resource requires `curl`, but it is not installed
* timeout while trying to connect to a database resource
* need to obtain information only available to `root`, but not running with sudo
* the test relies on examining a configuration file, but it does not exist (for example, `shadow` and `/etc/shadow`)

This is considered distinct from running a resource on an unsupported platform, which is detect by inspec before the test is attempted, and reported as a `skip`.

#### Expected Behavior

We currently handle test failures in some places by throwing an exception, or calling a helper method that throws the exception.  This bubbles up to be treated as a failed test.

A similar approach could be taken for `unable` tests, in terms of throwing an exception via a helper.

Overall, consensus within the InSpec team was that this should probably be handled as a type of failure; if it were treated as a type of skip, it would likely be ignored.

Refs #3017 #3158 

### 🌍 InSpec and Platform Version

2.2.x

**Comment by wardwouts** [comment]
I agree that tests that cannot be correctly executed should result in some sort of failure. To me the current behaviour where if I run this simple test, as a non-root user:

```
control "Tests reading in protected file" do
  title "Not allowed to read"
  desc "A file that cannot be read should not let a test succeed."
  describe file('/etc/shadow') do
    its('content') { should_not match /testvalue/ }
  end
end
```


the result is a pass, is obviously wrong.

**Comment by ghost** [comment]
I also agree that tests that cannot be correctly executed should result in some sort of a failure. Having test that did not even run report as succesfull is not the way to go.

**Comment by majormoses** [comment]
This caused a production issue that was meant to be caught by the tests I wrote, if the tests can't run it should report back so via an exit code indicating all is not well. There are maybe use cases where they should be skipped because the tests are flakey themselves but not something as obvious I dont have the deps to run the check.

**Comment by aaronlippold** [comment]
So are we actively working the above list ?

**Comment by majormoses** [comment]
Well I tried to open a pull request to start working on it but it was closed out for further discussion.

**Comment by majormoses** [comment]
Is there any progress on being able to move forward on this? I am happy to help but don't want to keep opening pull requests if we want to wait for more feedback. I think the overwhelming majority supports this behavior change and without it inspec is fundamentally not reliable for test automation or compliance as you can't trust the tests.

**Comment by clintoncwolfe** [comment]
Hi!  We really appreciate your interest in this, and I want to assure you that we on the InSpec engineering team consider this a critical issue.  Fixing it is important, but we must be sure to fix it in a way that is deliberate and well-considered.  In the two weeks since @majormoses commented on this, we've been very busy with two major other events: the release of InSpec 3.0, and (as I write this) the Seattle Community Summit. In the march to 3.0, we've been feature-focused; we plan for our time after 3.0 to circle back to be fix-focused.

Again we thank you for your patience, and any constructive ideas you have about how to handle the many small ways in which a resource can fail would be very welcome on this issue.

**Comment by majormoses** [comment]
I think addressing this prior to a major release would have made a lot of sense as even though this is a bug, it is a breaking (behavior) change and in an ideal world should happen in a major release.

> any constructive ideas you have about how to handle the many small ways in which a resource can fail would be very welcome on this issue.

What specifically are you looking for? It seems pretty straightforward to me, outside of platform not supported which is detected by inspec any test that fails to execute (be in lack of dependencies, permissions, connection timeouts, etc) should all result in a failed test. Are you looking for a complete list?

**Comment by clintoncwolfe** [comment]
Refs https://github.com/inspec/inspec-gcp/issues/61 , refs https://github.com/inspec/inspec/pull/3473

**Comment by clintoncwolfe** [comment]
FYI this is now being actively worked; we are hoping to make the 3.0 release, at least with the majority of the problem

**Comment by majormoses** [comment]
@clintoncwolfe should I start opening pull requests for some of the resources?

**Comment by clintoncwolfe** [comment]
Work was started at https://github.com/inspec/inspec/tree/cw/handle-unable . 

We made it as far as reworking how the `file` resource handles matchers and properties when the file does not exist.  Under 2.3.23 (master), for example, this will result in a PASS:

```
describe('/no/such/file') do
  its('content') { should be_nil }   # Arguable
  it { should_not be_readable }      # Misleading
  it { should_not be_readable.by_user('bob') } # Very misleading
end
```

Arguably, all those should be a FAIL (due to unable) You may or not agree that's in scope for this change.  We could call this "Greater Unable".  We'd need to do similar work on every resource.

Certainly, where things throw a Skip due to an inability, those should become Fails - most obvious example being [here](https://github.com/inspec/inspec/blob/master/lib/utils/file_reader.rb), but there are other places as well: anywhere we throw a [ResourceSkipped in a resource](https://github.com/inspec/inspec/search?q=ResourceSkipped&unscoped_q=ResourceSkipped). Changing those to fails isn't a difficult task.

Finally, there are resources that have peculiar behavior of all kinds; and we need to examine them and make decisions in each case.  We could call this - throwing skips when we should throw fails - as "Lesser Unable".

While reviewing these approaches, we noticed that either scope would break many profiles in use - those that individuals have made, those that Chef contributes to as part of an open-source effort like dev-sec, and the proprietary profiles Chef maintains.  This is a major conversion effort.

We've issued no deprecation warnings on this new behavior yet. Our normal policy for breaking changes is to issue warnings on next release, "n", and make the behavior change on "n+1".

With this in mind, our plan now is to introduce the changes to the behavior of resources - i.e. throw an exception instead of silently continuing - gradually. In the first iteration, after 3.0 is released, we will start to introduce deprecation warnings to the file and command resources whenever they are used to interrogate a file or command that doesn't exist. In all places where we would have silently failed you will now receive a deprecation warning. The behavior does not yet change in this iteration.

We will then continue to target all other resources in InSpec that need additional work. Once we achieve coverage, you will have a deprecation warning for all resources, whose future behavior will change.

Our current timeline is to cover all these within the minor releases of InSpec 3.x. We are also considering a mode to activate the future behavior in InSpec 3.x for all these cases (almost like a preview or testmode for the future release). Finally, we expect these deprecations to take effect during the InSpec 4.0 release.

**Comment by majormoses** [comment]
I personally think all of these examples with `file` should fail as the `file` resource supports ` it { should_not exist }`. If something does not exist then that implies `(nil|null)` content, there is no real value in any other of these "negative" assertions for `file`. If someone wrote a test for `its('content') { should be_nil }` I interpret this as the file should exist but be `nil` as you might be using something such as `touch $FILE` to keep track of the last time an action was done by inspecting the timestamp.

To give a more clear example building on what was already said:
```
describe file('/that/should/not/exist') do
  it { should_not exist }
  # the rest of these tests bring no value and should probably be removed from any existing tests/profiles if people are doing that
  its('content') { should be_nil }
  it { should_not be_readable }
  it { should_not be_readable.by_user('bob') }
end

describe file('/that/should/exist/but/does/not') do
  it { should exist }
  # these assertions make sense to run and fail on if the file does not exist
  its('content') { should be_nil }
  it { should_not be_readable }
  it { should_not be_readable.by_user('bob') }
end
```

I went through all the instances in the search and all of them where we skip make sense to fail on.

**Comment by aaronlippold** [comment]
I'm glad to see this moving ahead.

EDIT: (clintoncwolfe): remove quoted comment

**Comment by jquick** [comment]
This would be a huge breaking change if we started erroring instead of skipping on file/comman issues. A lot of profiles would be redone to accommodate it.

**Comment by frezbo** [comment]
@jquick Agree a lot of profiles would need to be re-written, but using Inspec as a compliance scanning tool requires the test to fail if the file does not exist, or if something fails the test should error out, rather than skipping, this could be well converted to an RFC.

**Comment by majormoses** [comment]
I agree that it will break things but this is a bug fix not a feature enhancement/request/change. Testing and more specifically compliance is pass/fail world. Just because it's an upstream bug does not change the fact that its fundamentally flawed. We may choose to version it as a major version to avoid people complaining but that does not change the fact that the tests/profiles that are written are currently broken because of this they just don't know it.

**Comment by majormoses** [comment]
Any progress to report? Is this being worked on? tagging @jerryaldrichiii  as we spoke about this at chefconf.

**Comment by majormoses** [comment]
spoke with @jerryaldrichiii and tagging @kekaichinose as requested.

**Comment by kekaichinose** [comment]
@majormoses apologies, I'm still coming up to speed on the mass of issues we have in the project to prioritize. I'll catch up with the team to understand this more and get back to you.

**Comment by majormoses** [comment]
please leave open

**Comment by majormoses** [comment]
> This issue has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. We value your input and contribution. Please leave a comment if this issue still affects you.

please leave open
